_posts/2021-04-22-torch-for-optimization/torch-for-optimization.Rmd

```diff
@@ -0,0 +1,640 @@
+$$
+
+, with $a$ and $b$ configurable parameters often set to 1 and 5, respectively.
+
+In R:
+
+```{r}
+library(torch)
+
+a <- 1
+b <- 5
+
+rosenbrock <- function(x) {
+  x1 <- x[1]
+  x2 <- x[2]
+  (a - x1)^2 + b * (x2 - x1^2)^2
+}
+```
+
+Its minimum is located at (1,1), inside a narrow valley surrounded by breakneck-steep cliffs:[^1]
+
+[^1]: The code used to generate this plot is found in the [appendix](#rosenbrock-function-plotting-code).
+
+```{r, eval = TRUE, echo = FALSE, layout="l-page", fig.cap="Rosenbrock function."}
+knitr::include_graphics("images/rosenbrock.png")
+```
+
+Our goal and strategy are as follows.
+
+We want to find the values $x_1$ and $x_2$ for which the function attains its minimum. We have to start somewhere; and from wherever that gets us on the graph we follow the negative of the gradient "downwards", descending into regions of consecutively smaller function value.
+
+Concretely, in every iteration, we take the current $(x1,x2)$ point, compute the function value as well as the gradient, and subtract some fraction of the latter to arrive at a new $(x1,x2)$ candidate. This process goes on until we either reach the minimum -- the gradient is zero -- or improvement is below a chosen threshold.
+
+Here is the corresponding code. For no special reasons, we start at `(-1,1)` . The learning rate (the fraction of the gradient to subtract) needs some experimentation. (Try 0.1 and 0.001 to see its impact.)
+
+```{r}
+num_iterations <- 1000
+
+# fraction of the gradient to subtract 
+lr <- 0.01
+
+# function input (x1,x2)
+# this is the tensor w.r.t. which we'll have torch compute the gradient
+x_star <- torch_tensor(c(-1, 1), requires_grad = TRUE)
+
+for (i in 1:num_iterations) {
+
+  if (i %% 100 == 0) cat("Iteration: ", i, "\n")
+
+  # call function
+  value <- rosenbrock(x_star)
+  if (i %% 100 == 0) cat("Value is: ", as.numeric(value), "\n")
+
+  # compute gradient of value w.r.t. params
+  value$backward()
+  if (i %% 100 == 0) cat("Gradient is: ", as.matrix(x_star$grad), "\n\n")
+
+  # manual update
+  with_no_grad({
+    x_star$sub_(lr * x_star$grad)
+    x_star$grad$zero_()
+  })
+}
+```
+
+    Iteration:  100 
+    Value is:  0.3502924 
+    Gradient is:  -0.667685 -0.5771312 
+
+    Iteration:  200 
+    Value is:  0.07398106 
+    Gradient is:  -0.1603189 -0.2532476 
+
+    ...
+    ...
+
+    Iteration:  900 
+    Value is:  0.0001532408 
+    Gradient is:  -0.004811743 -0.009894371 
+
+    Iteration:  1000 
+    Value is:  6.962555e-05 
+    Gradient is:  -0.003222887 -0.006653666 
+
+While this works, it really serves to illustrate the principle only. With `torch` providing a bunch of proven optimization algorithms, there is no need for us to manually compute the candidate $\mathbf{x}$ values.
+
+## Function minimization with `torch` optimizers
+
+Instead, we let a `torch` optimizer update the candidate $\mathbf{x}$ for us. Habitually, our first try is *Adam*.
+
+### Adam (?)
```

Comment:
Is naming attempts a numerical-optimization/computing/technical thing? If not, at this point (I'm reading the post in order, obviously), I'm not sure why we're doing this.